src/hotspot/share/compiler/oopMap.cpp

```diff
@@ -297,8 +315,8 @@ void OopMapSet::all_do(const frame *fr, const RegisterMap *reg_map,
           continue;
         }
 #ifdef ASSERT
-        if ((((uintptr_t)loc & (sizeof(*loc)-1)) != 0) ||
-            !Universe::heap()->is_in_or_null(*loc)) {
+        if ((((uintptr_t)loc & (sizeof(oop)-1)) != 0) ||
+            !Universe::heap()->is_in_or_null((oop)NativeAccess<AS_NO_KEEPALIVE>::oop_load(&val))) {
```

Comment:
This NativeAccess call causes the troubles with Shenandoah GC. It triggers evacuation of the object referenced by the stack location during a safepoint, which we'd rather avoid. Why is it needed?

Reply:
We call frame::oops_do from the GC to make bad oops good. But the oops_do logic assers the oop is good *before* the closure is applied.
Every time I do something, I run i to issues with this assert. It seems like an invalid assumption to me, that oops are meant to always be good, even before they are fixed. I vote for removing this assert. It has caused a lot of pain. Any takers?

Reply:
I think the previous assert was intentionally weaker: `is_in` checks that argument points to a committed area in the heap, which can include the oops not yet fixed. It does not check if oop is valid as far as GC is concerned. So maybe reverting `NativeAccess::oop_load` is enough?

Reply:
It was checking is_in_or_null before and after. Our is_in checks that it is in committed memory, and that the pointer is not stale w.r.t. color, as that is very likely to be a bug. We reluctantly made it relaxed for our safepoints that flip colors. I think it was once again for this very same usual suspect assert. But now that this is concurrent, its memory can get concurrently freed.

I think this is the 7th time I try to make this assert happy. I run into it all the time. IMO, the underlying assumption of the assert, is wrong. This is an iterator used by GC to fix totally invalid stale pointers into valid pointers. You really can't expect that they are valid before fixing them. That just happened to be true for other GCs. It makes little sense to me. That is why my preferred solution is to remove the assert. I would not miss it.

We have had similar issues with the oop_iterate framework asserting oops are valid before the closure is applied. At least there, it is possible to opt out...

Would anyone miss the assert?

Reply:
FTR, this is the RFE to remove the oop verification from the oop_iterate framework:
https://bugs.openjdk.java.net/browse/JDK-8237363

I really would like to get rid of it, but got push back because it made GCs have to duplicate the effort to provide verification.

Reply:
I've also has problems with this assert in the past, and I think that the underlying assumption of the assert is wrong. I would not miss it.

Reply:
IMO it's ok to remove it.
However, it can be argued that is_in() should not check for 'in committed memory' but 'in reserved space for heap', IOW 'is this a pointer into our heap memory region?'.  Or maybe this would be CH::is_in_reserved() and we should change the assert to that?
Also, IMO we should not do anything that may trip barriers on oop-iterators or similar GC-only paths.

Reply:
Almost always, what you want to do, is to assert the validity of heap pointers, where they are expected to be valid (dereferenceable pointer to a not freed object). And then you actually do want to check that they are within the committed boundaries of the heap. Otherwise you have a bug. We don't want to make such reasonable assertions weaker, because we cater for asserts that want oops to look valid pretty much all the time, even when they are stale pointers from a previous cycle, pointing into freed memory (until it is fixed). And I would much rather cater for asserts with good underlying reasoning (the oop should actually be okay here or something is wrong), than to cater for asserts that seem to just have the wrong underlying reasoning, making everyone else suffer for it, and letting bugs slip past all the other asserts.

The is_in_reserved check no longer exists. And I don't think I want to bring it back from the dead for this assert. Because I don't think the assert is right. The GCs already have their own verification code for roots (and the heap), that does this job, but much better. So I don't think it makes sense for the actual shared code oop iterator to go in and put constraints on what an invalid (not yet fixed up) stale pointer should and should not look like, for all GC implementations. Let the GC's own verification code take care of that. It has a much better idea about what a stale pointer should and should not look like.

And I do agree that we should not use barriers in this kind of code path. I just didn't know what else to do here to make the assert happy. Sorry about that. Well now I know. I am going to delete it, unless anyone opposes that solution.

Reply:
It's ok for me.